gax/src/main/java/com/google/api/gax/batching/v2/BatcherImpl.java

```diff
@@ -228,4 +241,25 @@ boolean hasAnyThresholdReached() {
       return elementCounter >= elementThreshold || byteCounter >= bytesThreshold;
     }
   }
+
+  /**
+   * Executes the {@link #sendBatch()} on a periodic interval. Also, it cancels the ongoing
+   * scheduler in case {@link Batcher} has been Garbage collected.
+   */
+  private class PushCurrentOpenBatch implements Runnable {
```

Comment:
This has to be a static class, otherwise you retain an implicit reference to the batcher preventing it from getting GCed

Reply:
Please have a look at the `PushCurrentBatchRunnable`, I have added a test case as well to verify the future is canceled after Batcher is GCed.

Please let me know if this class can be improved.